ratpack-test/src/main/java/ratpack/test/http/internal/DefaultMultipartForm.java

```diff
@@ -0,0 +1,396 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ratpack.test.http.internal;
+
+import ratpack.test.http.MultipartFileSpec;
+import ratpack.test.http.MultipartFormSpec;
+
+import java.io.UnsupportedEncodingException;
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/*
+    As defined in RFC 1867 and 2388 (see https://tools.ietf.org/html/rfc1867)
+ */
+public class DefaultMultipartForm {
+
+  private final Data data;
+  private final Writer writer;
+  private final Map<String, String> _headers;
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  private DefaultMultipartForm(Data data) {
+    this.data = data;
+    this.writer = new Writer(data);
+    this._headers = new HashMap<>();
+    _headers.put("Content-Type", getContentType());
+  }
+
+  public String getBody() {
+    return writer.write();
+  }
+
+  public String getContentType() {
+    return String.format("multipart/form-data; boundary=%s", data.boundary);
+  }
+
+  public Map<String, String> getHeaders() {
+    return Collections.unmodifiableMap(_headers);
+  }
+
+  final static class Writer {
```

Comment:
Boundry is being passed around quite a lot, but it is immutable from `Data` and could probably be extracted in the constructor and made field in this class which would simplify a number of signatures.

I'm also not a fan of passing the `StringBuilder` instance around and would probably pass it in the constructor and make `write()` a single use method, but that's probably a matter of personal preference.